File: src/nostr/wallet_registry_test.rs
use super::*;
use crate::common::test_helpers::*;
use crate::nostr::Tag;
use std::sync::{Arc, Mutex};



#[test]
fn test_registry_sub_persistence() {
    futures::executor::block_on(async {
        let storage = MockStorage::new();
        let mut registry = WalletRegistry::new(storage, Limits::default());
        
        let filters = vec![Filter {
            authors: Some(vec!["alice".into()]),
            ..Default::default()
        }];
        
        registry.subscribe(1, "sub1".into(), filters).await.unwrap();
        
        let data = registry.storage.data.lock().unwrap();
        assert!(data.contains_key("conn:1"), "Storage should contain connection state");
        assert!(data.contains_key("pk:alice"), "Storage should contain pubkey index");
    });
}

#[test]
fn test_hibernation_contract() {
    futures::executor::block_on(async {
        let storage = MockStorage::new();
        let mut registry = WalletRegistry::new(storage, Limits::default());

        registry.subscribe(1, "sub1".into(), vec![Filter {
            authors: Some(vec!["alice".into()]),
            ..Default::default()
        }]).await.unwrap();

        let mut registry2 = simulate_hibernation(&registry.storage).await;

        let event = Event {
            id: "e1".into(),
            pubkey: "alice".into(),
            created_at: 1000,
            kind: 23194,
            tags: vec![],
            content: "".into(),
            sig: "sig".into(),
        };
        let responses = registry2.match_event(&event).await;
        assert!(responses.contains(&RegistryResponse::WakeUp(1)),
            "hibernated connection must produce WakeUp");
        assert!(responses.contains(&RegistryResponse::Send {
            recipient_id: 1, sub_id: "sub1".into()
        }), "hibernated connection must produce Send");
    });
}

#[test]
fn test_hibernation_info_survival() {
    futures::executor::block_on(async {
        let storage = MockStorage::new();
        let mut registry = WalletRegistry::new(storage, Limits::default());

        registry.subscribe(1, "sub1".into(), vec![Filter {
            authors: Some(vec!["alice".into()]),
            ..Default::default()
        }]).await.unwrap();

        let info = Event {
            id: "info1".into(),
            pubkey: "alice".into(),
            created_at: 1000,
            kind: 13194,
            tags: vec![],
            content: "wallet info".into(),
            sig: "sig".into(),
        };
        registry.cache_info(info.clone()).await;

        let mut registry2 = simulate_hibernation(&registry.storage).await;

        let retrieved = registry2.get_info("alice").await;
        assert!(retrieved.is_some(), "info event should survive hibernation");
        assert_eq!(retrieved.unwrap().id, "info1");

        let event = Event {
            id: "e1".into(),
            pubkey: "alice".into(),
            created_at: 1000,
            kind: 23194,
            tags: vec![],
            content: "".into(),
            sig: "sig".into(),
        };
        let responses = registry2.match_event(&event).await;
        assert!(responses.contains(&RegistryResponse::WakeUp(1)),
            "hibernated connection must still produce WakeUp with info");
        assert!(responses.contains(&RegistryResponse::Send {
            recipient_id: 1, sub_id: "sub1".into()
        }), "hibernated connection must still produce Send with info");
    });
}

#[test]
fn test_hibernation_subscribe_cycles() {
    futures::executor::block_on(async {
        let storage = MockStorage::new();
        let mut registry = WalletRegistry::new(storage, Limits::default());

        registry.subscribe(1, "sub1".into(), vec![Filter {
            authors: Some(vec!["alice".into()]),
            ..Default::default()
        }]).await.unwrap();

        registry.unsubscribe(1, "sub1".into()).await.unwrap();

        registry.subscribe(1, "sub1".into(), vec![Filter {
            authors: Some(vec!["bob".into()]),
            ..Default::default()
        }]).await.unwrap();

        let mut registry2 = simulate_hibernation(&registry.storage).await;

        let alice_event = Event {
            id: "e1".into(),
            pubkey: "alice".into(),
            created_at: 1000,
            kind: 23194,
            tags: vec![],
            content: "".into(),
            sig: "sig".into(),
        };
        let responses = registry2.match_event(&alice_event).await;
        assert!(!responses.contains(&RegistryResponse::Send { recipient_id: 1, sub_id: "sub1".into() }),
            "alice event should NOT match after re-subscribe to bob");

        let bob_event = Event {
            id: "e2".into(),
            pubkey: "bob".into(),
            created_at: 1000,
            kind: 23194,
            tags: vec![],
            content: "".into(),
            sig: "sig".into(),
        };
        let responses2 = registry2.match_event(&bob_event).await;
        assert!(responses2.contains(&RegistryResponse::WakeUp(1)),
            "bob event must produce WakeUp after re-subscribe");
        assert!(responses2.contains(&RegistryResponse::Send {
            recipient_id: 1, sub_id: "sub1".into()
        }), "bob event must produce Send after re-subscribe");
    });
}

#[test]
fn test_registry_match_routing() {
    futures::executor::block_on(async {
        let storage = MockStorage::new();
        let mut registry = WalletRegistry::new(storage, Limits::default());
        
        let wallet_pk = "wallet_pk";
        registry.subscribe(1, "sub1".into(), vec![Filter {
            p_tags: Some(vec![wallet_pk.into()]),
            ..Default::default()
        }]).await.unwrap();
        
        let event = Event {
            id: "event1".into(),
            pubkey: "app_pk".into(),
            created_at: 1000,
            kind: 23194,
            tags: vec![Tag::p(wallet_pk)],
            content: "test".into(),
            sig: "sig".into(),
        };
        
        let matches = registry.match_event(&event).await;
        assert!(matches.contains(&RegistryResponse::WakeUp(1)));
        assert!(matches.contains(&RegistryResponse::Send { recipient_id: 1, sub_id: "sub1".into() }));
    });
}

#[test]
fn test_registry_lazy_load() {
    futures::executor::block_on(async {
        let storage = MockStorage::new();
        let wallet_pk = "hibernated_pk";
        let conn_id = 42;
        
        let mut entries = HashMap::new();
        entries.insert(format!("pk:{}", wallet_pk), serde_json::json!(vec![conn_id]));
        entries.insert(format!("conn:{}", conn_id), serde_json::json!({
            "subscriptions": {
                "sub1": [{"#p": [wallet_pk]}]
            },
            "info_event": null
        }));
        storage.put_batch(entries).await.unwrap();
        
        let mut registry = WalletRegistry::new(storage, Limits::default());
        
        let event = Event {
            id: "event1".into(),
            pubkey: "app_pk".into(),
            created_at: 1000,
            kind: 23194,
            tags: vec![Tag::p(wallet_pk)],
            content: "test".into(),
            sig: "sig".into(),
        };
        
        let responses = registry.match_event(&event).await;
        
        assert!(responses.contains(&RegistryResponse::WakeUp(conn_id)));
        assert!(responses.contains(&RegistryResponse::Send { recipient_id: conn_id, sub_id: "sub1".into() }));
    });
}

#[test]
fn test_index_matching_grouped() {
    futures::executor::block_on(async {
        let storage = MockStorage::new();
        let mut registry = WalletRegistry::new(storage, Limits::default());

        registry.subscribe(1, "sub1".into(), vec![Filter {
            authors: Some(vec!["alice".into()]),
            ..Default::default()
        }]).await.unwrap();
        registry.subscribe(2, "sub1".into(), vec![Filter {
            authors: Some(vec!["alice".into()]),
            ..Default::default()
        }]).await.unwrap();

        let event_alice = Event {
            id: "id".into(),
            pubkey: "alice".into(),
            kind: 1,
            tags: vec![],
            content: "".into(),
            sig: "".into(),
            created_at: 1000,
        };
        let responses = registry.match_event(&event_alice).await;

        let grouped: Vec<_> = responses.iter()
            .filter_map(|r| match r {
                RegistryResponse::Send { recipient_id, sub_id } if sub_id == "sub1" => Some(*recipient_id),
                _ => None,
            })
            .collect();
        assert_eq!(grouped.len(), 1);
        assert!(grouped.contains(&2));
        assert!(!grouped.contains(&1));
    });
}

#[test]
fn test_info_event_caching() {
    futures::executor::block_on(async {
        let storage = MockStorage::new();
        let mut registry = WalletRegistry::new(storage, Limits::default());

        let event = Event {
            id: "id1".into(),
            pubkey: "alice".into(),
            kind: 13194,
            tags: vec![],
            content: "".into(),
            sig: "sig1".into(),
            created_at: 1000,
        };
        registry.cache_info(event.clone()).await;

        let stored = registry.get_info("alice").await;
        assert!(stored.is_some());
        assert_eq!(stored.unwrap().id, event.id);
    });
}

#[test]
fn test_info_id_index() {
    futures::executor::block_on(async {
        let storage = MockStorage::new();
        let mut registry = WalletRegistry::new(storage, Limits::default());

        let event = Event {
            id: "id1".into(),
            pubkey: "alice".into(),
            kind: 13194,
            tags: vec![],
            content: "".into(),
            sig: "sig1".into(),
            created_at: 1000,
        };
        registry.cache_info(event.clone()).await;

        let pk = registry.find_info_pubkey_by_id("id1");
        assert_eq!(pk, Some("alice".into()));

        let pk_missing = registry.find_info_pubkey_by_id("nonexistent");
        assert_eq!(pk_missing, None);
    });
}

#[test]
fn test_delete_info() {
    futures::executor::block_on(async {
        let storage = MockStorage::new();
        let mut registry = WalletRegistry::new(storage, Limits::default());

        let event = Event {
            id: "id1".into(),
            pubkey: "alice".into(),
            kind: 13194,
            tags: vec![],
            content: "".into(),
            sig: "sig1".into(),
            created_at: 1000,
        };
        registry.cache_info(event.clone()).await;

        assert!(registry.get_info("alice").await.is_some());
        assert_eq!(registry.find_info_pubkey_by_id("id1"), Some("alice".into()));

        registry.delete_info("alice").await;

        assert!(registry.get_info("alice").await.is_none());
        assert_eq!(registry.find_info_pubkey_by_id("id1"), None);

        let data = registry.storage.data.lock().unwrap();
        assert!(!data.contains_key("info:alice"));
    });
}

#[test]
fn test_delete_info_preserves_subscriptions() {
    futures::executor::block_on(async {
        let storage = MockStorage::new();
        let mut registry = WalletRegistry::new(storage, Limits::default());

        registry.subscribe(1, "sub1".into(), vec![Filter {
            authors: Some(vec!["alice".into()]),
            ..Default::default()
        }]).await.unwrap();

        let event = Event {
            id: "id1".into(),
            pubkey: "alice".into(),
            kind: 13194,
            tags: vec![],
            content: "".into(),
            sig: "sig1".into(),
            created_at: 1000,
        };
        registry.cache_info(event.clone()).await;

        registry.delete_info("alice").await;

        assert!(registry.get_info("alice").await.is_none());
        assert!(registry.has_subscription(1, "sub1"));
    });
}

#[test]
fn test_registry_sync() {
    futures::executor::block_on(async {
        let storage = MockStorage::new();
        let mut registry = WalletRegistry::new(storage, Limits::default());

        registry.subscribe(1, "sub1".into(), vec![Filter {
            authors: Some(vec!["alice".into()]),
            ..Default::default()
        }]).await.unwrap();

        let data = registry.storage.data.lock().unwrap();
        assert!(data.contains_key("conn:1"));
        assert!(data.contains_key("pk:alice"));
    });
}

#[test]
fn test_registry_terminate() {
    futures::executor::block_on(async {
        let storage = MockStorage::new();
        let mut registry = WalletRegistry::new(storage, Limits::default());

        registry.subscribe(1, "sub1".into(), vec![Filter {
            authors: Some(vec!["alice".into()]),
            ..Default::default()
        }]).await.unwrap();

        registry.on_terminate(1).await;

        let data = registry.storage.data.lock().unwrap();
        assert!(!data.contains_key("conn:1"));
        assert!(!data.contains_key("pk:alice"));
        assert!(!registry.has_subscription(1, "sub1"));
    });
}

#[test]
fn test_registry_lazy_deletion() {
    futures::executor::block_on(async {
        let storage = MockStorage::new();
        let mut entries = HashMap::new();
        entries.insert("pk:stale".to_string(), serde_json::json!(vec![99]));
        storage.put_batch(entries).await.unwrap();

        let mut registry = WalletRegistry::new(storage, Limits::default());

        let result = registry.load_by_pubkey("stale").await;
        assert!(result.is_empty());

        let data = registry.storage.data.lock().unwrap();
        assert!(!data.contains_key("pk:stale"));
    });
}

#[test]
fn test_subscription_limit_exceeded() {
    futures::executor::block_on(async {
        let storage = MockStorage::new();
        let limits = Limits::new(65536, 0);
        let mut registry = WalletRegistry::new(storage, limits);

        let result = registry.subscribe(1, "sub1".into(), vec![Filter {
            authors: Some(vec!["alice".into()]),
            ..Default::default()
        }]).await;
        assert!(result.is_err());
        assert!(result.unwrap_err().to_string().contains("too many subscriptions"));
    });
}

#[test]
fn test_subscribe_rejected_on_storage_failure() {
    futures::executor::block_on(async {
        let storage = MockStorage {
            data: Arc::new(Mutex::new(HashMap::new())),
            fail_put_batch: true,
        };
        let mut registry = WalletRegistry::new(storage, Limits::default());

        let result = registry.subscribe(1, "sub1".into(), vec![Filter {
            authors: Some(vec!["alice".into()]),
            ..Default::default()
        }]).await;

        assert!(result.is_err());
        let err = result.unwrap_err().to_string();
        assert!(err.contains("persist failed"), "expected persist failure, got: {}", err);
        assert!(err.contains("mock storage unavailable"), "expected mock error message, got: {}", err);
    });
}

#[test]
fn test_unsubscribe_graceful_on_storage_failure() {
    futures::executor::block_on(async {
        let storage = MockStorage::new();
        let mut registry = WalletRegistry::new(storage, Limits::default());

        registry.subscribe(1, "sub1".into(), vec![Filter {
            authors: Some(vec!["alice".into()]),
            ..Default::default()
        }]).await.unwrap();
        registry.subscribe(1, "sub2".into(), vec![Filter {
            authors: Some(vec!["alice".into()]),
            ..Default::default()
        }]).await.unwrap();

        registry.storage.fail_put_batch = true;

        let result = registry.unsubscribe(1, "sub1".into()).await;
        assert!(result.is_ok(), "unsubscribe should not propagate storage error");

        assert!(!registry.has_subscription(1, "sub1"), "sub1 should be removed in-memory");
        assert!(registry.has_subscription(1, "sub2"), "sub2 should still exist");
    });
}

#[test]
fn test_delete_info_for_nonexistent_pubkey() {
    futures::executor::block_on(async {
        let storage = MockStorage::new();
        let mut registry = WalletRegistry::new(storage, Limits::default());
        registry.delete_info("nobody").await;
        assert!(registry.get_info("nobody").await.is_none());
    });
}

#[test]
fn test_read_pk_list_with_number_value() {
    futures::executor::block_on(async {
        let storage = MockStorage::new();
        let mut entries = HashMap::new();
        entries.insert("pk:single".to_string(), serde_json::json!(42));
        entries.insert("conn:42".to_string(), serde_json::json!({
            "subscriptions": {
                "sub1": [{"#p": ["single"]}]
            },
            "info_event": null
        }));
        storage.put_batch(entries).await.unwrap();

        let mut registry = WalletRegistry::new(storage, Limits::default());
        let ids = registry.load_by_pubkey("single").await;
        assert!(ids.contains(&42));
    });
}

#[test]
fn test_read_pk_list_with_unknown_value_type() {
    futures::executor::block_on(async {
        let storage = MockStorage::new();
        let mut entries = HashMap::new();
        entries.insert("pk:weird".to_string(), serde_json::json!("string_value"));
        storage.put_batch(entries).await.unwrap();

        let mut registry = WalletRegistry::new(storage, Limits::default());
        let ids = registry.load_by_pubkey("weird").await;
        assert!(ids.is_empty());
    });
}

#[test]
fn test_two_connections_same_pubkey_different_filters() {
    futures::executor::block_on(async {
        let storage = MockStorage::new();
        let mut registry = WalletRegistry::new(storage, Limits::default());

        let shared_pk = "shared_pk";

        registry.subscribe(1, "wallet_sub".into(), vec![Filter {
            kinds: Some(vec![23194]),
            p_tags: Some(vec![shared_pk.into()]),
            ..Default::default()
        }]).await.unwrap();

        registry.subscribe(2, "app_sub".into(), vec![Filter {
            kinds: Some(vec![23195]),
            authors: Some(vec![shared_pk.into()]),
            ..Default::default()
        }]).await.unwrap();

        #[allow(clippy::await_holding_lock)]
        let snapshot: HashMap<String, Value> = {
            let data = registry.storage.data.lock().unwrap();
            let pk_val = data.get(&format!("pk:{}", shared_pk)).expect("pk: entry should exist");
            let ids: Vec<u32> = serde_json::from_value(pk_val.clone()).unwrap();
            assert!(ids.contains(&1), "pk: entry should contain conn 1");
            assert!(ids.contains(&2), "pk: entry should contain conn 2");
            data.iter().map(|(k, v)| (k.clone(), v.clone())).collect()
        };

        {
            let storage2 = MockStorage::new();
            storage2.put_batch(snapshot.clone()).await.unwrap();
            let mut registry2 = WalletRegistry::new(storage2, Limits::default());

            let req_event = Event {
                id: "req1".into(),
                pubkey: "app_pk".into(),
                created_at: 1000,
                kind: 23194,
                tags: vec![Tag::p(shared_pk)],
                content: "pay invoice".into(),
                sig: "sig".into(),
            };
            let responses = registry2.match_event(&req_event).await;
            assert!(responses.contains(&RegistryResponse::WakeUp(1)));
            assert!(responses.contains(&RegistryResponse::Send { recipient_id: 1, sub_id: "wallet_sub".into() }));
            assert!(!responses.contains(&RegistryResponse::Send { recipient_id: 2, sub_id: "app_sub".into() }));
        }

        {
            let storage3 = MockStorage::new();
            storage3.put_batch(snapshot).await.unwrap();
            let mut registry3 = WalletRegistry::new(storage3, Limits::default());

            let resp_event = Event {
                id: "resp1".into(),
                pubkey: shared_pk.into(),
                created_at: 1000,
                kind: 23195,
                tags: vec![Tag::p("app_pk")],
                content: "paid".into(),
                sig: "sig".into(),
            };
            let responses2 = registry3.match_event(&resp_event).await;
            assert!(responses2.contains(&RegistryResponse::WakeUp(2)));
            assert!(responses2.contains(&RegistryResponse::Send { recipient_id: 2, sub_id: "app_sub".into() }));
            assert!(!responses2.contains(&RegistryResponse::Send { recipient_id: 1, sub_id: "wallet_sub".into() }));
        }
    });
}

File: src/nostr/wallet-registry.test.ts
import { describe, it, expect } from 'vitest';
import { MockStorage, seedSubscription } from '../common/test-helpers.ts';
import { DEFAULT_LIMITS, createLimits } from './limits.ts';
import { WalletRegistry } from './wallet-registry.ts';
import type { Event } from './event.ts';
import type { Filter } from './filter.ts';
import { Tag } from './tag.ts';

function hasSubscription(registry: WalletRegistry<MockStorage>, connId: number, subId: string): boolean {
  return registry.getConnSubscriptions(connId).has(subId);
}

function makeEvent(overrides: Partial<Event> = {}): Event {
  return {
    id: 'e1',
    pubkey: 'alice',
    createdAt: 1000,
    kind: 23194,
    tags: [],
    content: '',
    sig: 'sig',
    ...overrides,
  };
}

describe('WalletRegistry', () => {

  describe('subscription persistence', () => {
    it('stores subscription state in storage', async () => {
      const storage = new MockStorage();
      const registry = new WalletRegistry(storage, DEFAULT_LIMITS);

      const filters: Filter[] = [{ authors: ['alice'] }];
      await registry.subscribe(1, 'sub1', filters);

      expect(storage.data.has('conn:1')).toBe(true);
      expect(storage.data.has('pk:alice')).toBe(true);
    });
  });

  describe('hibernation contract', () => {
    function simulateHibernation(original: MockStorage): WalletRegistry<MockStorage> {
      const newStorage = new MockStorage();
      for (const [k, v] of original.data) {
        newStorage.data.set(k, v);
      }
      return new WalletRegistry(newStorage, DEFAULT_LIMITS);
    }

    it('produces WakeUp and Send after hibernation', async () => {
      const storage = new MockStorage();
      const registry = new WalletRegistry(storage, DEFAULT_LIMITS);

      await registry.subscribe(1, 'sub1', [{ authors: ['alice'] }]);

      const registry2 = simulateHibernation(storage);

      const event = makeEvent();
      const responses = await registry2.matchEvent(event);
      expect(responses).toContainEqual({ kind: 'wakeUp', connectionId: 1 });
      expect(responses).toContainEqual({ kind: 'send', recipientId: 1, subId: 'sub1' });
    });

    it('info event survives hibernation', async () => {
      const storage = new MockStorage();
      const registry = new WalletRegistry(storage, DEFAULT_LIMITS);

      await registry.subscribe(1, 'sub1', [{ authors: ['alice'] }]);

      const info = makeEvent({
        id: 'info1',
        kind: 13194,
        content: 'wallet info',
      });
      await registry.cacheInfo(info);

      const registry2 = simulateHibernation(storage);

      const retrieved = await registry2.getInfo('alice');
      expect(retrieved).not.toBeNull();
      expect(retrieved!.id).toBe('info1');

      const event = makeEvent();
      const responses = await registry2.matchEvent(event);
      expect(responses).toContainEqual({ kind: 'wakeUp', connectionId: 1 });
      expect(responses).toContainEqual({ kind: 'send', recipientId: 1, subId: 'sub1' });
    });

    it('unsubscribe and re-subscribe updates correctly after hibernation', async () => {
      const storage = new MockStorage();
      const registry = new WalletRegistry(storage, DEFAULT_LIMITS);

      await registry.subscribe(1, 'sub1', [{ authors: ['alice'] }]);
      await registry.unsubscribe(1, 'sub1');
      await registry.subscribe(1, 'sub1', [{ authors: ['bob'] }]);

      const registry2 = simulateHibernation(storage);

      const aliceEvent = makeEvent({ pubkey: 'alice', id: 'e1' });
      const aliceResponses = await registry2.matchEvent(aliceEvent);
      expect(aliceResponses).not.toContainEqual({ kind: 'send', recipientId: 1, subId: 'sub1' });

      const bobEvent = makeEvent({ pubkey: 'bob', id: 'e2' });
      const bobResponses = await registry2.matchEvent(bobEvent);
      expect(bobResponses).toContainEqual({ kind: 'wakeUp', connectionId: 1 });
      expect(bobResponses).toContainEqual({ kind: 'send', recipientId: 1, subId: 'sub1' });
    });
  });

  describe('event matching', () => {
    it('routes events to matching subscriptions', async () => {
      const storage = new MockStorage();
      const registry = new WalletRegistry(storage, DEFAULT_LIMITS);

      const walletPk = 'wallet_pk';
      await registry.subscribe(1, 'sub1', [{ pTags: [walletPk] }]);

      const event = makeEvent({
        id: 'event1',
        pubkey: 'app_pk',
        tags: [Tag.p(walletPk)],
        content: 'test',
      });
      const matches = await registry.matchEvent(event);
      expect(matches).toContainEqual({ kind: 'wakeUp', connectionId: 1 });
      expect(matches).toContainEqual({ kind: 'send', recipientId: 1, subId: 'sub1' });
    });

    it('lazy loads from storage on match', async () => {
      const storage = new MockStorage();
      const walletPk = 'hibernated_pk';
      const connId = 42;

      await seedSubscription(storage, connId, 'sub1', walletPk, [{ pTags: [walletPk] }]);

      const registry = new WalletRegistry(storage, DEFAULT_LIMITS);

      const event = makeEvent({
        id: 'event1',
        pubkey: 'app_pk',
        tags: [Tag.p(walletPk)],
        content: 'test',
      });

      const responses = await registry.matchEvent(event);
      expect(responses).toContainEqual({ kind: 'wakeUp', connectionId: connId });
      expect(responses).toContainEqual({ kind: 'send', recipientId: connId, subId: 'sub1' });
    });

    it('groups matching connections by latest ID', async () => {
      const storage = new MockStorage();
      const registry = new WalletRegistry(storage, DEFAULT_LIMITS);

      await registry.subscribe(1, 'sub1', [{ authors: ['alice'] }]);
      await registry.subscribe(2, 'sub1', [{ authors: ['alice'] }]);

      const event = makeEvent({ id: 'id', kind: 1 });
      const responses = await registry.matchEvent(event);

      const sends = responses.filter(
        (r): r is { kind: 'send'; recipientId: number; subId: string } =>
          r.kind === 'send' && r.subId === 'sub1',
      );
      expect(sends.length).toBe(1);
      expect(sends[0].recipientId).toBe(2);
    });
  });

  describe('info event caching', () => {
    it('caches and retrieves info events', async () => {
      const storage = new MockStorage();
      const registry = new WalletRegistry(storage, DEFAULT_LIMITS);

      const event = makeEvent({ id: 'id1', kind: 13194, sig: 'sig1' });
      await registry.cacheInfo(event);

      const stored = await registry.getInfo('alice');
      expect(stored).not.toBeNull();
      expect(stored!.id).toBe('id1');
    });

    it('indexes info events by ID', async () => {
      const storage = new MockStorage();
      const registry = new WalletRegistry(storage, DEFAULT_LIMITS);

      const event = makeEvent({ id: 'id1', kind: 13194, sig: 'sig1' });
      await registry.cacheInfo(event);

      expect(registry.findInfoPubkeyById('id1')).toBe('alice');
      expect(registry.findInfoPubkeyById('nonexistent')).toBeNull();
    });

    it('deletes info events', async () => {
      const storage = new MockStorage();
      const registry = new WalletRegistry(storage, DEFAULT_LIMITS);

      const event = makeEvent({ id: 'id1', kind: 13194, sig: 'sig1' });
      await registry.cacheInfo(event);

      expect(await registry.getInfo('alice')).not.toBeNull();
      expect(registry.findInfoPubkeyById('id1')).toBe('alice');

      await registry.deleteInfo('alice');

      expect(await registry.getInfo('alice')).toBeNull();
      expect(registry.findInfoPubkeyById('id1')).toBeNull();
      expect(storage.data.has('info:alice')).toBe(false);
    });

    it('preserves subscriptions when deleting info', async () => {
      const storage = new MockStorage();
      const registry = new WalletRegistry(storage, DEFAULT_LIMITS);

      await registry.subscribe(1, 'sub1', [{ authors: ['alice'] }]);

      const event = makeEvent({ id: 'id1', kind: 13194, sig: 'sig1' });
      await registry.cacheInfo(event);

      await registry.deleteInfo('alice');

      expect(await registry.getInfo('alice')).toBeNull();
      expect(hasSubscription(registry, 1, 'sub1')).toBe(true);
    });

    it('deleteInfo for nonexistent pubkey is a no-op', async () => {
      const storage = new MockStorage();
      const registry = new WalletRegistry(storage, DEFAULT_LIMITS);

      await registry.deleteInfo('nobody');
      expect(await registry.getInfo('nobody')).toBeNull();
    });
  });

  describe('sync and terminate', () => {
    it('sync persists subscription state', async () => {
      const storage = new MockStorage();
      const registry = new WalletRegistry(storage, DEFAULT_LIMITS);

      await registry.subscribe(1, 'sub1', [{ authors: ['alice'] }]);

      expect(storage.data.has('conn:1')).toBe(true);
      expect(storage.data.has('pk:alice')).toBe(true);
    });

    it('terminate removes all state', async () => {
      const storage = new MockStorage();
      const registry = new WalletRegistry(storage, DEFAULT_LIMITS);

      await registry.subscribe(1, 'sub1', [{ authors: ['alice'] }]);
      await registry.onTerminate(1);

      expect(storage.data.has('conn:1')).toBe(false);
      expect(storage.data.has('pk:alice')).toBe(false);
      expect(hasSubscription(registry, 1, 'sub1')).toBe(false);
    });

    it('lazy deletion removes stale pubkey entries', async () => {
      const storage = new MockStorage();
      await storage.putBatch({ 'pk:stale': [99] });

      const registry = new WalletRegistry(storage, DEFAULT_LIMITS);
      const result = await registry.loadByPubkey('stale');
      expect(result).toEqual([]);
      expect(storage.data.has('pk:stale')).toBe(false);
    });
  });

  describe('limits and error handling', () => {
    it('rejects subscriptions exceeding the limit', async () => {
      const storage = new MockStorage();
      const registry = new WalletRegistry(storage, createLimits(65536, 0));

      await expect(
        registry.subscribe(1, 'sub1', [{ authors: ['alice'] }]),
      ).rejects.toThrow('too many subscriptions');
    });

    it('rejects subscribe on storage failure', async () => {
      const storage = new MockStorage();
      storage.failPutBatch = true;
      const registry = new WalletRegistry(storage, DEFAULT_LIMITS);

      await expect(
        registry.subscribe(1, 'sub1', [{ authors: ['alice'] }]),
      ).rejects.toThrow('persist failed');
    });

    it('unsubscribe is graceful on storage failure', async () => {
      const storage = new MockStorage();
      const registry = new WalletRegistry(storage, DEFAULT_LIMITS);

      await registry.subscribe(1, 'sub1', [{ authors: ['alice'] }]);
      await registry.subscribe(1, 'sub2', [{ authors: ['alice'] }]);

      storage.failPutBatch = true;

      await registry.unsubscribe(1, 'sub1');

      expect(hasSubscription(registry, 1, 'sub1')).toBe(false);
      expect(hasSubscription(registry, 1, 'sub2')).toBe(true);
    });
  });

  describe('pk list edge cases', () => {
    it('reads pk list stored as a number', async () => {
      const storage = new MockStorage();
      await storage.putBatch({
        'pk:single': 42,
        'conn:42': {
          subscriptions: { sub1: [{ '#p': ['single'] } as unknown as Filter] },
          info_event: null,
        },
      });

      const registry = new WalletRegistry(storage, DEFAULT_LIMITS);
      const ids = await registry.loadByPubkey('single');
      expect(ids).toContain(42);
    });

    it('handles unknown pk list value type', async () => {
      const storage = new MockStorage();
      await storage.putBatch({ 'pk:weird': 'string_value' });

      const registry = new WalletRegistry(storage, DEFAULT_LIMITS);
      const ids = await registry.loadByPubkey('weird');
      expect(ids).toEqual([]);
    });
  });

  describe('two connections same pubkey', () => {
    it('routes events to correct connections', async () => {
      const storage = new MockStorage();
      const registry = new WalletRegistry(storage, DEFAULT_LIMITS);

      const sharedPk = 'shared_pk';

      await registry.subscribe(1, 'wallet_sub', [{ kinds: [23194], pTags: [sharedPk] }]);
      await registry.subscribe(2, 'app_sub', [{ kinds: [23195], authors: [sharedPk] }]);

      // Check pk entry contains both connections
      const pkVal = storage.data.get(`pk:${sharedPk}`);
      expect(Array.isArray(pkVal)).toBe(true);
      expect(pkVal).toContain(1);
      expect(pkVal).toContain(2);

      // Snapshot and create new registry
      const snapshot = new Map(storage.data);
      {
        const storage2 = new MockStorage();
        for (const [k, v] of snapshot) storage2.data.set(k, v);
        const registry2 = new WalletRegistry(storage2, DEFAULT_LIMITS);

        const reqEvent = makeEvent({
          id: 'req1',
          pubkey: 'app_pk',
          tags: [Tag.p(sharedPk)],
          content: 'pay invoice',
        });
        const responses = await registry2.matchEvent(reqEvent);
        expect(responses).toContainEqual({ kind: 'wakeUp', connectionId: 1 });
        expect(responses).toContainEqual({ kind: 'send', recipientId: 1, subId: 'wallet_sub' });
        expect(responses).not.toContainEqual({ kind: 'send', recipientId: 2, subId: 'app_sub' });
      }

      {
        const storage3 = new MockStorage();
        for (const [k, v] of snapshot) storage3.data.set(k, v);
        const registry3 = new WalletRegistry(storage3, DEFAULT_LIMITS);

        const respEvent = makeEvent({
          id: 'resp1',
          pubkey: sharedPk,
          kind: 23195,
          tags: [Tag.p('app_pk')],
          content: 'paid',
        });
        const responses2 = await registry3.matchEvent(respEvent);
        expect(responses2).toContainEqual({ kind: 'wakeUp', connectionId: 2 });
        expect(responses2).toContainEqual({ kind: 'send', recipientId: 2, subId: 'app_sub' });
        expect(responses2).not.toContainEqual({ kind: 'send', recipientId: 1, subId: 'wallet_sub' });
      }
    });
  });
});
